### DISCUSSION TITLE
Error loading ASGI app. Could not import module "src.main"

### DISCUSSION THREAD

**Discussion by Bilalharoon** [fastapi/fastapi#6600] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Depends, FastAPI, HTTPException, status
from fastapi.security import OAuth2PasswordBearer, OAuth2PasswordRequestForm
from fastapi.security.oauth2 import OAuth2PasswordRequestForm
from . import models
from . import services
from .database import engine, get_db
from passlib.context import CryptContext
from sqlalchemy.orm import Session
from datetime import datetime, timedelta

app = FastAPI()

oauth2_scheme = OAuth2PasswordBearer(tokenUrl='login')

pwd_context = CryptContext(schemes=['bcrypt'], depracated='auto')

models.Base.metadata.create_all(bind=engine)

SECRET_KEY = '[SECRET KEY]'
ALGORITHM = 'HS256'
ACCESS_TOKEN_EXPIRE_MINUTES = 30



@app.get("/")
async def root():
    return {"message": "Hello World"}
    
@app.get("/test/")
async def test(token:str = Depends(oauth2_scheme)):
    return {'token', token}

@app.post('/login/{method}')
async def login(method: str, form_data: OAuth2PasswordRequestForm = Depends(), db: Session = Depends(get_db), ):
    user = services.authenticate_user(form_data.username, form_data.password)
    if not user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Incorrect username or password",
            headers={"WWW-Authenticate": "Bearer"},
        )
    access_token_expires = timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
    access_token = services.create_access_token(
        data={"username": user.username, "email": user.email}, expires_delta=access_token_expires
    )
    return {"access_token": access_token, "token_type": "bearer"}
```

### Description

I saw the solution is changing 'main' -> 'src.main' so I tried that, but the same problem is occuring Even if I use 'cd src' to move to src and run 'uvicorn main:app --reload', errors continue. What do I need?

### Operating System

Linux

### Operating System Details

WSL 2

### FastAPI Version

0.68.1

### Python Version

3.8.10

**Comment by ghandic** [comment, marked as answer]
Can you give us your directory structure and the directory you are in when you run your uvicorn command?

You should be in the same directory as your main file to run `uvicorn main:app` so it will go to the main.py file and find the app object

**Comment by Bilalharoon** [comment]
> Can you give us your directory structure and the directory you are in when you run your uvicorn command?
> 
> You should be in the same directory as your main file to run `uvicorn main:app` so it will go to the main.py file and find the app object

![image](https://user-images.githubusercontent.com/25674730/135541526-2cd340c6-0073-447b-8e64-9dd77555a997.png)

**Comment by ghandic** [comment]
Where are you running your command from? what is the directory? Could you run `pwd` and an `ls`

**Comment by Bilalharoon** [comment]
pwd:
`/home/bilalharoon/Projects/Kalaam/kalaam-fastapi/src`

ls from inside src
`__init__.py  __pycache__  database.py  main.py  models.py  schemas.py  services.py`

ls from outside src
`alembic  kalaam.db  requirements.txt  src  venv`

**Comment by ghandic** [comment]
and `uvicorn main:app --reload` isnt working? Can you share the console output?

**Comment by Bilalharoon** [comment]
```
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [314] using watchgod
ERROR:    Error loading ASGI app. Could not import module "src.main".
```

**Comment by ghandic** [comment]
Looks like you're running `uvicorn src.main:app --reload` instead of `uvicorn main:app --reload`

**Comment by Bilalharoon** [comment]
I tried both. both of them don't work

**Comment by ghandic** [comment]
Can you confirm by running the whole sequence? It should be picking it up if the entrypoint you are running from is correct

```bash
pwd
ls -l
uvicorn main:app --reload
```

Please add the output so we can help - this isn't a fastapi issue, it is the way you're running your code.

**Comment by Bilalharoon** [comment]
So I made it work and I have no idea how. But what I did was change all the imports to module imports.

so instead of 
```python
import database
```
or
```python
from . import database
```

I did:
```python
from src import database
```
or
```python
import src.database
```

**Comment by peterbb148** [comment]
Have exact same problem. Here's my code:

```
from fastapi import FastAPI

app = FastAPI()

@app.get("/")
async def root():
    return {"message": "Hello World"}
```

Command line:

```
uvicorn fastapi:app --reload --port 8001
INFO:     Will watch for changes in these directories: ['/Users/peter/Git/peterbb148/Python']
INFO:     Uvicorn running on http://127.0.0.1:8001 (Press CTRL+C to quit)
INFO:     Started reloader process [3942] using statreload
ERROR:    Error loading ASGI app. Could not import module "fastapi".
```

All file are in the same directory.

**Comment by jmanteigueiro** [comment]
@peterbb148 what is the name of the file where that code is located?

**Comment by mikkokotila** [comment]
This should be re-opened. The issue can arise for something different that is covered in this ticket in terms of solutions. 

The error message is misleading in the sense that it can appear for other reasons than "not being able to find module". 

In my case the culprit is import in one of the submodules with `from app import dictionary_lookup` which is exactly the way in which all the other sub-modules handle the imports. For some reason, doing this in a particular file causes the fastapi error.

EDIT: It turns out to be simple case of having imports in wrong order. 

Given how many people are dealing with this issues, and many threads end up with no resolution, probably a good time to give ability to have verbosity, or clear the language of the error.

**Comment by samnimoh** [comment]
importing FastAPI before import uvicorn fixed it for me

**Comment by magal-ti** [comment]
just add a empty file in the same folder with this name _init_.py

**Comment by Amet-Vikram** [comment]
I just encountered the exact same problem. I noticed the error resolves as soon as I comment out certain imports. ``

```
from fastapi import FastAPI, Response, status, HTTPException, Depends, APIRouter
from fastapi.params import Body
import psycopg2
from psycopg2.extras import RealDictCursor
import time
from . import models, utils, Schemas
from .database import engine, get_db, SessionLocal
from sqlalchemy.orm import Session
from .routers import post2, user2 
```
The server starts again after I remove the last import. 

This is the command I am using: 


`uvicorn Apps.main:app --reload `

These are the from post2 (same for user2) directories from routers.

```
from fastapi import FastAPI, Response, status, HTTPException, Depends, APIRouter
from sqlalchemy.orm import Session
from ..database import get_db
from .. import models, utils, Schemas
```

But I do require to use the APIRouter which is the main reason for separating all the path operations from main file.

**Comment by aj2622** [comment]
change the name of your file and try again

**Reply by balogandras** [reply]
Thanks, it worked for me, but I have no idea why.
main.py - error
main_.py - works

**Reply by u1vi** [reply]
It works for me too, but 
the original file name was: media.py
First I changed the file to the  t_media.py (I got the same error)
Then changed it to api.py and it works.

It would be better if the fast API team return informative error message

**Comment by filipinjo** [comment]
The error message will appear if there is circular import error. Unfortunately, instead of circular import error the “Error loading ASGI app. Could not import module ….” is reported.

**Comment by zulfiqar-ali01** [comment]
the same error occurs in Pycharm cmd with the correct src folder

INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [12072] using statreload
WARNING:  The --reload flag should not be used in production on Windows.
ERROR:    Error loading ASGI app. Could not import module "app".

**Reply by Mehrnaz4Hosseini** [reply]
First find the directory of your python file. (main.py)
If you are using macOS you can get the path by clicking on get info. After that write this command:
cd Path/To/Your/Python/File/Directory

then again run: uvicorn main:app --reload

it solved my problem.

**Comment by ahmedlab8** [comment]
I had the same issue and it happened without any change from myside. 
I simply solved it by just creating another python script(copied my file from the app folder) in the same virtual environment folder.

**Comment by karlosdaniel451** [comment]
> The error message will appear if there is circular import error. Unfortunately, instead of circular import error the “Error loading ASGI app. Could not import module ….” is reported.

I got the mentioned error message but was able to solve this problem after solving a circular import error I had.

**Comment by michaldev** [comment]
I have this same problem. Here is simple example:
https://github.com/michaldev/problem-with-fastapi-import

I tried multiple versions:
- relative/absolute imports 
- uvicorn app.main:app and uvicorn main:app and cd app && uvicorn main:app

I can't import app object.

**Comment by SsayanJ** [comment]
I had the same issue with a very minimal example:
```python
from fastapi import FastAPI

app = FastAPI()

@app.get('/')
async def root():
    return {"message": "Basic API"}
```

Since everyone seems to be pointing to import issues, I added an unused import before as I had in previous FastAPI work I've done and it works... Can't really understand what is happening here but with the below it works:
```python
import pandas as pd   #Only modification
from fastapi import FastAPI

app = FastAPI()

@app.get('/')
async def root():
    return {"message": "Basic API"}
```

**Comment by JuliaBars** [comment]
I got absolutely the same issue, Error loading ASGI app. Attribute "app" not found in module "main". 
While it was like this, it didn't work:
```
from fastapi import FastAPi
app = FastAPI()
```

but since I've changed the import, it started:
```
import fastapi

app = fastapi.FastAPI()
```

just in case, someone else will get the same issue. Windows 11, home edition.

**Comment by omar-haddad99** [comment]
In your main.py file, import uvicorn (after import fastapi)
